### PR TITLE
Refactor of Rig and MyAvatar animation JavaScript interface

### PR DESCRIPTION
**Pull request by hyperlogic**
Large refactoring of the Rig class and it's consumers.  The goal here is to remove complexity by dropping support for the old animation system and to normalize/document the various coordinate spaces used.  Also, this brings full JavaScript support to the new animation system. 

Menu option changes:
* Removed Develper > Avatar > Enable Rig Animations
* Removed Develper > Avatar > Enable Anim Graph
* Added Developer > Avatar > Debug Draw Position
* Changed Develper > Avatar > Debug Draw Bind Pose to Debug Draw Default Pose
* Removed Develper > Avatar > Show IK Constraints

JavaScript changes:

* coordinate system of MyAvatar.setJoint* has changed, it is now directly
  in object coordinates, no longer in pre-rotated space.
* coordinate system of AnimVars used for IK targets are in Rig space
* removed MyAvatar.playAnimation
* removed MyAvatar.stopAnimation
* removed MyAvatar.getGetAnimationDetails
* removed MyAvatar.startAnimationByRole
* removed MyAvatar.stopAnimationByRole
* removed MyAvatar.getGetAnimationDetailsByRole
* removed MyAvatar.clearJointPriorities
* added MyAvatar.overrideAnimation(url, fps, loop, firstFrame, lastFrame)
* added MyAvatar.restoreAnimation()
* added MyAvatar.getAnimationRoles()
* added MyAvatar.overrideRoleAnimation(role, url, fps, loop, firstFrame, lastFrame)
* added MyAvatar.restoreRoleAnimation(role)
* added MyAvatar.prefetchAnimation(url)
* added two examples: kneel.js and theBird.js

C++ changes:

* Removal of AnimationHandle class
* Added getParent() and replaceChild() to AnimNode
* Added findByName() and traverse() to AnimNode
* Changed AnimStateMachine to hold nodes by childIndex instead of smart pointer. This allows script to replace nodes dynamically via overrideRoleAnimation
* Added accessors for default poses to AnimSkeleton and Rig.
* Rig implements script per-joint overrides using big flags and a set of override AnimPoses.
* Rig now contains methods to copy animation data into and out of the JointData structures.
* JointStates have been removed
* Removed EntityRig and AvatarRig
* One Frame Lag in Debug Draw has been fixed

Known blockers that need to be fixed before this PR is ready for merge:

Nice to have fixes:

  * Squatting while wearing an HMD can sometimes cause the feet to penetrate the floor, then pop into a squat.
  * Refactor/rename JavaScript set joint API, I prefer the override/restore names over set and clear.
  * Remove methods with priority arguments
  * cull dead/code all unused set/gets
  * optimize buildAbsolutePoses

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: db21db3cbcffffce823b763353d92c82780c513f available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by ZappoMan** [comment]
How much testing have you done with animated entities? We probably need to consider any break in backward compatibility for that feature as a merge blocker. In other words, animated entities that used to properly animate, need to still animate.

**Comment by howard-stearns** [comment]
Agreed @ZappoMan. One such test is walk.js. (It turns out that the current walk.js is already broken in master, but we've been working with DaveDub to get a new version that works in master and will work in this branch, too.) 

Do you have a suggested specific test for entity animation (e.g, a script)? We'll make sure it stays working.

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -1082,8 +944,8 @@ glm::vec3 MyAvatar::getDefaultEyePosition() const {
     return getPosition() + getWorldAlignedOrientation() * _skeletonModel.getDefaultEyeModelPosition();
 }
 
-const float SCRIPT_PRIORITY = DEFAULT_PRIORITY + 1.0f;
-const float RECORDER_PRIORITY = SCRIPT_PRIORITY + 1.0f;
+const float SCRIPT_PRIORITY = 1.0f + 1.0f;
+const float RECORDER_PRIORITY = 1.0f + 1.0f;
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: **SCRIPT_PRIORITY** and **RECORDER_PRIORITY** now appear to be cruft.  I believe they can be removed.

**Comment by hyperlogic** [comment]
@ZappoMan @howard-stearns  I've added that to the blocker list.   I have some test models, and currently, the animated Model entity rotations are in the wrong frame.  Will fix

File: libraries/animation/src/AnimNode.h [hunk 2]
```diff
@@ -78,6 +79,30 @@ class AnimNode {
...
+        Pointer result;
+        traverse([&](Pointer node) {
+            if (id == node->getID()) {
+                result = node;
+                return false;
+            }
+            return true;
+        });
+        return result;
+    }
```
Comment (howard-stearns) [inline libraries/animation/src/AnimNode.h]: Note used? I have no objection to keeping stuff that might be useful going forward -- I've wanted this one myself once -- but we don't normally do so and I want to make sure it's intentional.

File: libraries/animation/src/AnimNode.h [hunk 3]
```diff
@@ -78,6 +79,30 @@ class AnimNode {
...
+    bool traverse(F func) {
+        if (func(shared_from_this())) {
+            for (auto&& child : _children) {
+                if (!child->traverse(func)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
```
Comment (howard-stearns) [inline libraries/animation/src/AnimNode.h]: Not used except by findByName, below.

**Comment by AndrewMeadows** [comment]
I noticed that there are a few other features that might be temporarily hacked/disabled/broken:

* eye tracking
* bounding capsule of avatar replaced with a default capsule

What are your plans for fixing these: pre or post merge?  Would you like other devs to checkout your branch and help reimplement these kinds of hacks?

File: libraries/animation/src/Rig.h [hunk 4]
```diff
@@ -91,162 +63,184 @@ class Rig : public QObject, public std::enable_shared_from_this<Rig> {
...
-    int indexOfJoint(const QString& jointName);
-
-    void initJointTransforms(glm::mat4 rootTransform);
-    void clearJointTransformTranslation(int jointIndex);
-    void reset(const QVector<FBXJoint>& fbxJoints);
+
+    void overrideAnimation(const QString& url, float fps, bool loop, float firstFrame, float lastFrame);
+    void restoreAnimation();
+    QStringList getAnimationRoles() const;
+    void overrideRoleAnimation(const QString& role, const QString& url, float fps, bool loop, float firstFrame, float lastFrame);
```
Comment (howard-stearns) [inline libraries/animation/src/Rig.h]: Where is this implemented?

File: libraries/animation/src/AnimNode.h (hunk 3 above)
Comment (hyperlogic) [inline libraries/animation/src/AnimNode.h]: It is used, this is how we replace animations by role.

See Rig::void Rig::overrideRoleAnimation()

Comment (hyperlogic) [inline libraries/animation/src/AnimNode.h]: Sorry, I miss understood, travers is used to enumerate all animations by role.  See Rig::getAnimationRoles()

**Comment by hyperlogic** [comment]
@AndrewMeadows  Yes bounds and eyes are broken, also see the list of other blocker issues in the bottom of the description.  My plan is to fix these, then we can take off the Do Not Merge tag

File: libraries/animation/src/AnimNode.h (hunk 2 above)
Comment (howard-stearns) [inline libraries/animation/src/AnimNode.h]: Never mind. Diffs truncated by github.

**Comment by hifi-gustavo** [comment]
Build console output for commit: dd3f43705c9d5232c1038af08a3c8e0168f1271e available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8f46b8a765525b90a6d0a300cf9de7497b4a8e47 available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by howard-stearns** [comment]
Congratulations! Separate avatar threading appears to work! (Mac PR build 3329.) Specifically, mesh, various skeleton drawing, and reset all seems to function while threaded. (The latter is particularly cool, because recent master fails in AnimationHandle::simulate() when you hold the reset (apostrophe) key down while threaded.)  N.B.: I still need to try multiple avatars and check judder.

However: trying to turn avatar threading back off goes into an infinite loop!

**Comment by hifi-gustavo** [comment]
Build console output for commit: 14189ac90950cbc8b0c7143127c25a7a57972693 available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by howard-stearns** [comment]
Animation of ScriptableAvatars appears to work. (e.g., using examples/acScripts/animatedAvatarAgent.js)

**Comment by hifi-gustavo** [comment]
Build console output for commit: f120e10ff4ee9739aa47c3985bb109dca7408c4d available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 76034d62d578e6c442fbd9e8fd7155200e2bd807 available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 527199bc6919b2b45e86dd2b4d1b7d9222cd66bf available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 868cf83b4819458a95e9a28b5fa2df51de57b83d available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 97a2eb62d4771e08765a9c5bdac66cd38a86e23d available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2a9adafc099815e73ae8938100703197c08d0ce2 available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5072339a2241eb9405586a5bfa46ea4b23b8b68d available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7c9cb9778389ad1f44535350a98fd28e42ce6619 available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 549a8f7130dc8fc31c476049c4d16306fa9e2710 available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by howard-stearns** [comment]
I thought we were getting rid of of the meat-hook? In any case, should we be reporting when avatars don't have the expected parts (i.e., when we're falling back to defaults that might be wrong)?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 11bfed4185be7639f1506505ee9fa44e2fb25f2a available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by howard-stearns** [comment]
Enable avatar threading reliably locks up interface.

**Comment by hifi-gustavo** [comment]
Build console output for commit: e2d9e37eb886046bf236b505dce227f05bc67a5f available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f922f61c3dfc0bdfdf9dc843c7099e0c089013e8 available at: https://gist.github.com/05180a18658b470711ff<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 31cecdc174af3bbd110992b46bde6cc2b8ce134b available at: https://gist.github.com/c551a4eddc4cf8ab67e3<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6461.dmg

**Comment by howard-stearns** [comment]
Separate threading doesn't work well with this PR. Rather than wait, we're just removing the menu item  that turns it on.

**Pull request merged**